### PR TITLE
adds flatmap datapipe.

### LINKED ISSUE #178: Add a flatmap datapipe
### 🚀 The feature

Add a `FlatMapDataPipe` which simulataneously flattens nested `IterDataPipe`s and applies a function to nested pipes.

### Motivation, pitch

I have tarballs containing tarballs. When dealing with these types of recursive data, it's useful to have a mechanism to both flatten the structure and apply a function to it so I only need to consider the "inner-most" data.

### Alternatives

Domain libraries can define their own impls.

### Additional context

Migrating torchtext to datapipes via torchdata would benefit from this feature when dealing with [IWSLT data](https://github.com/pytorch/text/blob/main/torchtext/datasets/iwslt2016.py#L231-L241).

### PR DESCRIPTION
**Pull request by erip**
Fixes #178

### Changes

- Adds a `FlatMapDataPipe` which applies a function to each nested data pipe and yields from them to flatten lazily.

### CODE REVIEW

File: torchdata/datapipes/iter/transform/flatmap.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+from torch.utils.data import DataChunk
+from torch.utils.data import functional_datapipe, IterDataPipe
+
+from typing import TypeVar
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+@functional_datapipe("flatmap")
+class FlatMapDataPipe(IterDataPipe[IterDataPipe[DataChunk[T_co]]]):
```
Comment (ejguan) [inline torchdata/datapipes/iter/transform/flatmap.py:11]: Could you change the name to `FlatMapperIterDataPipe`? We want to name each DataPipe class as `op-er` and functional API as the `op`.

File: torchdata/datapipes/iter/transform/flatmap.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        datapipe: Iterable datapipe containing iterable datapipes to which the function is applied
+        fn: the function to be applied to each of the ``inner" datapipes
+    """
+    def __init__(self, datapipe, fn):
+        self.datapipe = datapipe
+        self.fn = fn
+
+    def __iter__(self):
+        for e in self.datapipe:
+            yield from e.map(self.fn)
```
Comment (ejguan) [inline torchdata/datapipes/iter/transform/flatmap.py:28]: Let's not assume each data inside the `self.datapipe` is an `IterDataPipe`. We can do:
```suggestion
        for e in self.datapipe:
            yield from self.fn(e)
```

File: torchdata/datapipes/iter/transform/flatmap.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    Iterable DataPipe which applies a function to an IterableDataPipe containing
+    IterableDataPipes and flattens to a single unnested IterableDataPipe.
+
+    Args:
+        datapipe: Iterable datapipe containing iterable datapipes to which the function is applied
+        fn: the function to be applied to each of the ``inner" datapipes
+    """
+    def __init__(self, datapipe, fn):
+        self.datapipe = datapipe
+        self.fn = fn
```
Comment (ejguan) [inline torchdata/datapipes/iter/transform/flatmap.py:24]: Could you also add a check lambda function here? https://github.com/pytorch/pytorch/blob/b372be4211d40be92ffbb025cac0038ee3445141/torch/utils/data/datapipes/iter/callable.py#L47

`lambda` function can not be pickled.

File: test/test_datapipe.py [hunk 4]
```diff
@@ -569,6 +570,14 @@ def _return_self(x):
         # __len__ Test: returns the number of batches
         self.assertEqual(3, len(batch_dp))
 
+    def test_flatmap_datapipe(self):
```
Comment (ejguan) [inline test/test_datapipe.py:573]: Could you please add a more comprehensive tests?
We at least need to add validation on `reset` using `reset_after_n_next_calls` shown from the other test cases, and `__len__` test.

**Review by ejguan — COMMENTED** [review]
Thanks for adding this PR, we do want to implement `flatmap`.

Could you please also take a reference from `MapperIterDataPipe` here. https://github.com/pytorch/pytorch/blob/b372be4211d40be92ffbb025cac0038ee3445141/torch/utils/data/datapipes/iter/callable.py#L105-L131

Since we want to make sure each DataPipe instance is serializable for the future snapshotting, could you please add the similar `__getstate__` and `__setstate__` to make sure `self.fn` is serialized if `dill` is available.

File: test/test_datapipe.py (hunk 4 above)
Comment (erip) [reply test/test_datapipe.py:573]: Yes, I'll add some tests momentarily.

**Comment by erip** [comment]
I've made some changes to the test. One that I'm a bit sketchy on is the `__len__` test since the length is a function of the size of the output of the `fn` (explained in-line in the test).

As an aside, I can't seem to test this locally? I've installed pytorch nightly and see this error when I run `pytest test/test_datapipe.py`: 

> ImportError: cannot import name 'DILL_AVAILABLE' from 'torch.utils.data.datapipes.utils.common'

**Comment by ejguan** [comment]
> I've made some changes to the test. One that I'm a bit sketchy on is the `__len__` test since the length is a function of the size of the output of the `fn` (explained in-line in the test).

Then, we should expect `__len__` not implemented in this case. It's allowed for `IterDataPipe`. And, a test to validate an Error would be raised would be sufficient.

> As an aside, I can't seem to test this locally? I've installed pytorch nightly and see this error when I run `pytest test/test_datapipe.py`:
> 
> > ImportError: cannot import name 'DILL_AVAILABLE' from 'torch.utils.data.datapipes.utils.common'
> 
> It seems like CI only runs for certain people, so I don't know if this is OK 😞

This is due to PyTorch nightly release for both mac and windows has been stalled for 10 days... 😢

**Comment by erip** [comment]
That's fine! So I should remove the __len__ test and we just need to wait?

**Comment by ejguan** [comment]
> That's fine! So I should remove the `__len__` test and we just need to wait?

You can do `self.assertRaises` for `__len__` test.

**Comment by erip** [comment]
OK, great!

File: torchdata/datapipes/iter/__init__.py [hunk 5]
```diff
@@ -41,6 +41,7 @@
     OnlineReaderIterDataPipe as OnlineReader,
 )
 from torchdata.datapipes.iter.transform.bucketbatcher import BucketBatcherIterDataPipe as BucketBatcher
+from torchdata.datapipes.iter.transform.flatmap import FlatMapDataPipe as FlatMap
```
Comment (ejguan) [inline torchdata/datapipes/iter/__init__.py:44]: nit: name needs to be changed here.

File: test/test_datapipe.py [hunk 6]
```diff
@@ -569,6 +570,27 @@ def _return_self(x):
...
+        source_dp = IterableWrapper(list(range(20)))
+
+        def fn(e):
+            return [e, e*10]
+
+        flatmapped_dp = nested_wrapper.flatmap(fn)
```
Comment (ejguan) [inline test/test_datapipe.py:579]: nit: nested_wrapper -> source_dp

**Review by ejguan — APPROVED** [review]
The PyTorch nightly is updated. Could you please address the lint error? Then, I can import the PR.

**Comment by erip** [comment]
OK, @ejguan -- I can run tests locally now and it seems good now.

File: test/test_datapipe.py [hunk 7]
```diff
@@ -569,6 +570,27 @@ def _return_self(x):
...
+        list_flatmapped_dp = list(flatmapped_dp)
+        expected_list = list(itertools.chain([(e, e*10) for e in source_dp]))
```
Comment (ejguan) [inline test/test_datapipe.py:581]: Should we do another `self.assertEqual(list_flatmapped_dp, expected_list)` here?

File: torchdata/datapipes/iter/transform/flatmap.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+from torch.utils.data import DataChunk
+from torch.utils.data import functional_datapipe, IterDataPipe
+from torch.utils.data.datapipes.utils.common import DILL_AVAILABLE, check_lambda_fn
+
+from typing import TypeVar
+
+if DILL_AVAILABLE:
+    import dill
```
Comment (ejguan) [inline torchdata/datapipes/iter/transform/flatmap.py:9]: The Error from `mypy` can be ignored here. I have another PR that should solve it.

**Comment by ejguan** [comment]
Could you please do a rebase onto main branch?

**Comment by erip** [comment]
Rebased!

File: test/test_datapipe.py [hunk 9]
```diff
@@ -1,9 +1,11 @@
 # Copyright (c) Facebook, Inc. and its affiliates.
 import io
+import itertools
 import unittest
 import warnings
 
 from collections import defaultdict
+from functools import partial
```
Comment (ejguan) [inline test/test_datapipe.py:8]: Last nit: Could you please also remove `partial` here, as it's not used?

Comment (erip) [reply test/test_datapipe.py]: Done! Not sure how linting missed that, sorry.

**Comment by erip** [comment]
@ejguan I can't see what the Meta-internal issues are, but I'm happy to try to resolve any warnings if it makes sense for me to do that. Otherwise, I'm excited to see this land. 😎

**Comment by ejguan** [comment]
> @ejguan I can't see what the Meta-internal issues are, but I'm happy to try to resolve any warnings if it makes sense for me to do that. Otherwise, I'm excited to see this land. 😎

No worries. I did check these Error and they are not related to this PR. I have triggered shipment of this PR and it will be merged shortly.
Thanks a lot!! @erip

**Pull request closed without merging**